swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,927 @@
+use std::sync::Arc;
+use std::task::{Context, Poll, Waker};
+use std::time::{Duration, Instant};
+use void::Void;
+
+/// A low-level building block for protocols that can be expressed as async functions.
+///
+/// An async function in Rust is executed within an executor and thus only has access to its local state
+/// and by extension, the state that was available when the [`Future`] was constructed.
+///
+/// This [`ConnectionHandler`] aims to reduce the boilerplate for protocols which can be expressed as
+/// a static sequence of reads and writes to a socket where the response can be generated with limited
+/// "local" knowledge or in other words, the local state within the [`Future`].
+///
+/// For outbound substreams, arbitrary data can be supplied via [`InEvent::NewOutbound`] which will be
+/// made available to the callback once the stream is fully negotiated.
+///
+/// Inbound substreams may be opened at any time by the remote. To facilitate this one and more usecases,
+/// the supplied callbacks for inbound and outbound substream are given access to the handler's `state`
+/// field. State can be shared between the [`NetworkBehaviour`] and a [`FromFn`] [`ConnectionHandler`]
+/// via the [`Shared`] abstraction.
+///
+/// [`Shared`] tracks a piece of state and updates all registered [`ConnectionHandler`]s whenever the
+/// state changes.
+///
+/// [`NetworkBehaviour`]: crate::NetworkBehaviour
+pub fn from_fn(protocol: &'static str) -> Builder<WantState> {
+    Builder {
+        protocol,
+        phase: WantState {},
+    }
+}
+
+pub struct Builder<TPhase> {
+    protocol: &'static str,
+    phase: TPhase,
+}
+
+pub struct WantState {}
+
+pub struct WantInboundHandler<TState> {
+    state: Arc<TState>,
+}
+
+pub struct WantOutboundHandler<TState, TInbound> {
+    state: Arc<TState>,
+    max_inbound: usize,
+    inbound_handler: Option<
+        Box<
+            dyn Fn(
+                    NegotiatedSubstream,
+                    PeerId,
+                    ConnectedPoint,
+                    Arc<TState>,
+                ) -> BoxStream<'static, TInbound>
+                + Send,
+        >,
+    >,
+}
+
+impl Builder<WantState> {
+    pub fn with_state<TState>(
+        self,
+        shared: &Shared<TState>,
+    ) -> Builder<WantInboundHandler<TState>> {
+        Builder {
+            protocol: self.protocol,
+            phase: WantInboundHandler {
+                state: shared.shared.clone(),
+            },
+        }
+    }
+
+    pub fn without_state(self) -> Builder<WantInboundHandler<()>> {
+        Builder {
+            protocol: self.protocol,
+            phase: WantInboundHandler {
+                state: Arc::new(()),
+            },
+        }
+    }
+}
+
+impl<TState> Builder<WantInboundHandler<TState>> {
+    pub fn with_inbound_handler<TInbound, TInboundStream>(
+        self,
+        max_inbound: usize,
+        handler: impl Fn(NegotiatedSubstream, PeerId, ConnectedPoint, Arc<TState>) -> TInboundStream
```

Comment:
Please include the protocol name (also for outbound).

Reply:
At the moment, there is only one protocol possible so it will always be that one if that function is even called.

Reply:
I know, and I’d like to [change that](https://github.com/rkuhn/blog/blob/master/08_peer-to-peer_internet.md). All prefabricated upgrades today ignore the negotiated protocol name, which to my mind leaves the second most important value provided by libp2p on the table (the first one being the ability to dial a PeerId).